proposals/0082-relax-transaction-constraints.md

```diff
@@ -0,0 +1,173 @@
+---
+simd: '0082'
+title: Relax Transaction Constraints
+authors:
+  - Andrew Fitzgerald (Solana Labs)
+category: Standard
+type: Core
+status: Draft
+created: 2023-10-30
+feature:
+---
+
+## Summary
+
+This proposal aims to relax some of the constraints on which individual
+transactions can be included in a valid block.
+The proposal does not relax the constraints required for a transaction to be
+executed.
+
+## Motivation
+
+The current protocol places many constraints on the structure and contents
+of blocks; if any constraints are broken, block-validators will mark the block
+as invalid.
+Many of these constraints are necessary, but some of them are not, and lead to
+additional complexity in the protocol.
+This proposal aims to relax some of the constraints at the individual
+transaction level, in order to simplify the protocol, and give more flexibility
+to block-producer and block-validator implementations.
+
+More specifically, this proposal aims to relax the constraints which require
+account state in order to determine the validity of a transaction's inclusion.
+The reason these constraints are targetted specifically, is that reliance on
+account state necessarily requires synchronous execution within the protocol.
+This proposal on its' own, will not enable asynchronous execution, but it will
+remove one of the barriers to asynchronous execution.
+
+## Alternatives Considered
+
+1. Do nothing
+    - This is the simplest option, as we could leave the protocol as is.
+    However, this leaves the protocol more complex than it needs to be.
+2. Relax all but fee-paying constraint
+    - This was actually the initial proposal, but it was decided that it would
+    be better to relax as many constraints as possible, rather than just some.
+    Any reliance on account state, means the protocol requires synchronous
+    execution.
+3. Additionally, relax the address lookup table resolution constraint
+    - This was considered, since it is a transaction-level constraint that is
+    depdendent on account-state. However, due to entry-level and block-level
+    constraints that rely on the address lookup table resolution, this
+    constraint cannot easily be relaxed without also relaxing those
+    constraints.
+
+## New Terminology
+
+None
+
+## Detailed Design
+
+Specifically, this proposal relaxes constraints that a transaction included in
+a block must:
+
+1. Have a fee-payer with enough funds to pay fees
+2. Have a valid nonce account, if specified
```

Comment:
Did you consider that both normal transactions and durable nonce transactions can only be processed once to be valid? Validators and bankless leaders can use the status cache to avoid including normal transactions more than once, but they need to check that the nonce is valid or not to avoid including durable nonce transactions more than once.

Reply:
Considered, but I need to rewrite sections of the proposal to be very explicit.

Status cache should still be used to avoid including transactions multiple times. Duplicate inclusion of transactions is still a violation of constraints and leads to the entire block being marked as invalid. This does not depend on account state, which is why I think this should be left in.

With respect to nonces, we want to avoid depdendence on account state, so cannot check the validity of the nonce account for the question of block validity. However, the nonce acount state must still be checked before execution. If the nonce account does not exist or is otherwise invalid, the transaction is not executed and fees are not charged to the fee-payer.
I view this as a similar situation to the fee-payer check; The block-producer is incentivized to check these states. They are free to do that in some non-strict method, but if done incorrectly, they lose out on rewards and the transaction has no/limited on state and have effectively burned their CUs.

There's an ongoing discussion of if fees should be paid by the block-producer should a transaction be unable to pay. So whether or not the producer gets charged fees in these cases is still an open question.